src/utils/polygon.cpp

```diff
@@ -370,21 +399,63 @@ void PolygonRef::simplify(const coord_t smallest_line_segment_squared, const coo
         //h^2 = L^2 / b^2     [factor the divisor]
         const coord_t height_2 = area_removed_so_far * area_removed_so_far / base_length_2;
         if ((height_2 <= 1 //Almost exactly colinear (barring rounding errors).
-            && LinearAlg2D::getDist2FromLine(current, previous, next) <= 1) // make sure that height_2 is not small because of cancellation of positive and negative areas
-            || (length2 < smallest_line_segment_squared
-                && next_length2 < smallest_line_segment_squared // Segments are small
-                && height_2 <= allowed_error_distance_squared) // removing the vertex doesn't introduce too much error.
-        )
+            && LinearAlg2D::getDist2FromLine(current, previous, next) <= 1)) // make sure that height_2 is not small because of cancellation of positive and negative areas
         {
-            continue; //Remove the vertex.
+            continue;
         }
+  
+        if (length2 < smallest_line_segment_squared
+            && height_2 <= allowed_error_distance_squared) // removing the vertex doesn't introduce too much error.) // removing the vertex doesn't introduce too much error.
+        {
+            if(next_length2 > smallest_line_segment_squared)
+            {
+                // Special case; The next line is long. If we were to remove this, it could happen that we get quite noticeable artifacts.
+                // We should instead move this point to a location where both edges are kept and then remove the previous point that we wanted to keep.
+                // By taking the intersection of these two lines, we get a point that perseves the direction (so it makes the corner a bit more pointy)
+                Point intersection_point;
+                bool has_intersection = lineLineIntersection(previous_previous, previous, current, next, &intersection_point);
 
-        //Don't remove the vertex.
+                if(has_intersection)
+                {
+                    // Find out if it is a degenerate intersection.
+                    if(vSize2(current - intersection_point) > allowed_error_distance_squared)
+                    {
+                        // The intersection is way to far away. Drop it. 
+                        continue;
+                    }
+                    
+                    // New point seems like a valid one.
+                    current = intersection_point;
+                    
+                    // If there was a previous point added, remove it.
+                    if(new_path.size() > 0)
+                    {
+                        new_path.pop_back();
+                    }
+                    else
+                    {
+                        continue;
```

Comment:
Wait, I really don't think this should be a continue. Why alter current if you're gonna throw it away right after?
I think the continue should be omitted, so that the intersection_point gets added as the first thing on the output poly. Then at the end the algorithm should already omit the very last point, correct?

In any case this line of code needs a comment to justify / explain it.

Reply:
Note that this may violate the allowed error distance constraint, because here we only checked on `height_2 <= allowed_error_distance_squared` , but we didn't check on the hieght_2 of the previous vert.

![image](https://user-images.githubusercontent.com/8895761/92922218-c001aa00-f435-11ea-9eda-23dddb4e87d3.png)

Reply:
> Wait, I really don't think this should be a continue. Why alter current if you're gonna throw it away right after?
I think the continue should be omitted, so that the intersection_point gets added as the first thing on the output poly. Then at the end the algorithm should already omit the very last point, correct?

Agreed! As it stands now, `current` would get skipped instead of `previous` in this case, and you might alter a very long line segment between `current` and `next`.

Reply:
> Note that this may violate the allowed error distance constraint, because here we only checked on height_2 <= allowed_error_distance_squared , but we didn't check on the hieght_2 of the previous vert.

No, because we move `current` to the intersection between `prevprev -> prev` and `current -> next`. Forgive my crude edit of your image:
![simplify_intersection](https://user-images.githubusercontent.com/2448634/93226217-5bb35300-f773-11ea-85ba-7377833e96ac.png)
So that ignored height should be 0.

Reply:
Agreed. The `continue` problem is solved, so this is not an issue any more.